### PR TITLE
Complete Implementation of Cat

### PR DESCRIPTION
**Pull request by mmstick**
- Implemented caret and M- notation.
- Write directly to stdout rather than buffering to a vector.
- Integrate exit status with Program struct

### CODE REVIEW

File: src/bin/cat.rs [hunk 1]
```diff
@@ -197,13 +253,24 @@ impl Program {
                 }
             }
         }
-        std::process::exit(exit_status);
+        return self.exit_status.get();
```
Comment (Ticki) [inline src/bin/cat.rs]: You can remove this `return`.

File: src/bin/cat.rs [hunk 2]
```diff
@@ -197,13 +253,24 @@ impl Program {
...
+    stdout.write(&[b'^']).try(stderr);
+    stdout.write(&[notation]).try(stderr);
+}
+
 fn main() {
     let stdout = io::stdout();
     let mut stdout = stdout.lock();
     let mut stderr = io::stderr();
-    Program::initialize(&mut stdout, &mut stderr).execute(&mut stdout, &mut stderr);
+    std::process::exit(Program::initialize(&mut stdout, &mut stderr).and_execute(&mut stdout, &mut stderr));
```
Comment (Ticki) [inline src/bin/cat.rs]: Import the `process` module.

File: src/bin/cat.rs [hunk 3]
```diff
@@ -153,38 +156,91 @@ impl Program {
...
+                            127 => push_caret(stdout, stderr, b'?'),
+                            128...159 => {
+                                stdout.write(b"M-^").try(stderr);
+                                stdout.write(&[byte-64]).try(stderr);
+                            },
+                            160...255 => {
+                                stdout.write(b"M-").try(stderr);
+                                stdout.write(&[byte-128]).try(stderr);
+                            },
+                            _ => unreachable!() // u8 number range is from 0-255
```
Comment (Ticki) [inline src/bin/cat.rs]: You don't need this. You can simply make the arm above catch everything else.

**Comment by Ticki** [comment]
Left a few comments, otherwise LFTM. Awesome work.

**Comment by Ticki** [comment]
https://travis-ci.org/redox-os/coreutils/builds/122338577

**Comment by mmstick** [comment]
It's caused by recent commits made to master.

**Comment by mmstick** [comment]
I'll close this and re-open it.

**Pull request closed without merging**